### ISSUE TITLE
automate app deployment

### ISSUE THREAD

**Issue by simsong** [digitalcorpora/digitalcorpora_app#65]
Should we automate app deployment? I can set it up so that commits to bottle-mail are automatically deployed to downloads.digitalcorpora.org. This could be done by an hourly script that does a pull on the repo, or this could be done with an SSH deployment key that ssh'es into the VM every time there is a merge with bottle-mail and does the pull.

Any thoughts? @ajnelson-nist ? 

- [x] @ajnelson-nist  reviews deployment script
- [ ] Set up pull from github on timer or github push

**Comment by ajnelson-nist** [comment]
I suppose the question is how much faith you have in the unit tests guaranteeing we don't have site-wreckers make it into `bottle-main`.

If you believe the unit tests cover site liveness well enough, I'm fine with an automated pull being triggered contingent on CI passing.

**Comment by simsong** [comment]
They don't. They did with flask, but flask caused problems at Dreamhost.
On the other hand, the script could do the merge into a temp branch and then test the live site and then roll back if there is a problem.

**Comment by ajnelson-nist** [comment]
> They don't. They did with flask, but flask caused problems at Dreamhost. On the other hand, the script could do the merge into a temp branch and then test the live site and then roll back if there is a problem.

Is there a "Canary" subdomain that temp branch would display under?  That seems like the most tenable approach.

**Comment by simsong** [comment]
There are multiple domains:
* stats.
* corp.
* downloads.
* dev.


I do the live development under dev. But we've seen things work on dev and then not work on downloads because they have different python installations. Ideally, this would be a virtual env, and that virtual env would also be part of the git repo. But I just realized that advantage as I typed this. And virtual envs still depend upon the base installation...

**Comment by simsong** [comment]
There is a deployment script that I am now using. 
And we have more self-tests!
https://app.codecov.io/gh/digitalcorpora/digitalcorpora_app/tree/bottle-main

**Comment by ajnelson-nist** [comment]
Is the app deployment automatic?  Or do you still need to do something manual to deploy [PR 72](https://github.com/digitalcorpora/digitalcorpora_app/pull/72)?

**Comment by simsong** [comment]
I have to do something manually....

**Comment by simsong** [comment]
check now.

**Comment by ajnelson-nist** [comment]
Thank you, the state with PR 72's live.

Back to this thread: Eventually, once CI passes from `bottle-main`, I think it'd be reasonable to run that deploy script.  I say that not knowing what's in the script, though.

**Comment by simsong** [comment]
https://github.com/digitalcorpora/digitalcorpora_app/blob/bottle-main/etc/deploy.bash

**Comment by ajnelson-nist** [comment]
Ah, so I see now the reliance on the host's Python installation.

Would it work better with a virtual environment built as part of the Makefile recipes?  My usual style of handling that is, make the venv and a flag file for it, then have every recipe that needs the venv import the venv.  E.g.:

https://github.com/casework/CASE-Utilities-Python/blob/0.9.0/tests/Makefile#L85-L93

It's a little repetitive within the Makefiles, but removes virtual environment considerations from Bash-level script calls to Make targets.

**Comment by simsong** [comment]
I don't want to put the entire virtual environment into git. 
There is a local python environment. You just don't see it in that script.

**Comment by ajnelson-nist** [comment]
> I don't want to put the entire virtual environment into git.

Sorry, I didn't meant to suggest that.  I always ignore the virtual environment and flag file - they're built as part of CI.

If there's a persistent local Python environment: It looks like the deployment script potentially modifies it, depending on ...oh, nevermind, `requirements.txt` is currently not a Make-encoded dependency (looking at `install-dependencies`).

I have no complaints if this is working well enough.  I do suspect that shifting to a virtual environment within the Makefile can bypass some of those troubles you were seeing where different servers' Pythons were sufficiently different to be a problem---if you haven't gotten past that problem already, that is.

You can check off that "I reviewed it" box in the first message of the thread now if you'd like.

**Comment by simsong** [comment]
> You can check off that "I reviewed it" box in the first message of the thread now if you'd like.

You are the person who is supposed to check it off!

I'm thinking of having the deployment script first do a checkout into a junk directory and test it there. What do you think of that? Otherwise, I am thinking of going live with this. Question: should it poll, or should GitHub do a push?

**Comment by ajnelson-nist** [comment]
> > You can check off that "I reviewed it" box in the first message of the thread now if you'd like.
> 
> You are the person who is supposed to check it off!
> 

Done.

> I'm thinking of having the deployment script first do a checkout into a junk directory and test it there. What do you think of that?

I'm fine with this.  `mktemp`'s OK with me.

> Otherwise, I am thinking of going live with this. Question: should it poll, or should GitHub do a push?

My preference between the two is slightly more towards GitHub pushing.

**Comment by simsong** [comment]
> My preference between the two is slightly more towards GitHub pushing.

This means github has a ssh key to log into the account...

**Comment by ajnelson-nist** [comment]
GitHub could be given an SSH key for a secondary account that only gets a headless Git repository, or even just a shell that only gets to touch exactly one path.  The web user could then use `inotify` to trigger a pull, test, and build within the host environment.

Though, that would make SSH access issues a bear to debug.  Periodic `git pull`s from a `cron` job would be fine.  Is every ten minutes too aggressive?

**Comment by ajnelson-nist** [comment]
I'd thought you had a push-from-GitHub deployment workflow set up.  If you don't, `cron` is fine.

**Comment by simsong** [comment]
I'll investigate if github can ssh into the account and run the script.  The account is pretty locked down, and it only accesses the repo with https

**Comment by simsong** [comment]
Okay, I have a design.
https://stackoverflow.com/questions/60710209/trigger-github-actions-only-when-pr-is-merged

I'll get this implemented at some point, but not for a bit.

**Comment by simsong** [comment]
No longer an issue since we are now using Zappa and moving to SAM for Lambda.
